mobilecoind-dev-faucet/src/slam/mod.rs

```diff
@@ -0,0 +1,535 @@
+                    // slam
+                    return Err(format!("Cannot obtain more utxos, aborting slam: {}", err));
+                }
+            }
+        }
+
+        // Close this so that workers that get to the end of the queue won't block
+        prepared_utxos_sender.close();
+
+        // Get the network state, which contains block height, now so that it is
+        // fresh for the slam workers
+        let network_state = {
+            let resp = mobilecoind_api_client
+                .get_network_status(&Default::default())
+                .map_err(|err| format!("Failed getting network status: {}", err))?;
+            self.block_height
+                .store(resp.network_highest_block_index, Ordering::SeqCst);
+            resp
+        };
+
+        // Now, we spawn worker threads, which build txs and
+        // submits them, with retries, in parallel
+        self.phase
+            .store(SlamState::SUBMITTING_TRANSACTIONS, Ordering::SeqCst);
+        log::info!(logger, "Slam status: {}", self.get_status().unwrap());
+        let begin_submit_time = Instant::now();
+        let prepare_time = begin_submit_time.duration_since(begin_prepare_time);
+
+        // Note: Workers are std::thread rather than tokio::task because mc_connection
+        // currently only has blocking APIs If we get an async-api for
+        // attestation and tx submission, we could change this to be tasks and just use
+        // the tokio thread pool. maybe some advantages?
+        let workers: Vec<_> = (0..params.num_threads)
+            .map(|worker_num| {
+                let this = self.clone();
+                let params = params.clone();
+                let prepared_utxos_receiver = prepared_utxos_receiver.clone();
+                let recipient = recipient.clone();
+                let account_key = account_key.clone();
+                let network_state = network_state.clone();
+                let tx_submitter = tx_submitter.clone();
+                let logger = logger.new(o! { "thread" => format!("slam-worker-{}", worker_num) });
+                std::thread::spawn(move || {
+                    this.slam_worker_entry_point(
+                        worker_num,
+                        &params,
+                        prepared_utxos_receiver,
+                        recipient,
+                        account_key,
+                        network_state,
+                        tx_submitter,
+                        logger,
+                    )
+                })
+            })
+            .collect();
+
+        // Wait for each worker to finish
+        log::info!(logger, "Waiting for workers to finish");
+        for worker in workers {
+            // Note: this is blocking, we could use async_thread instead I guess. But it
+            // only blocks one thread at most since only one slam can happen at
+            // a time, so it likely doesn't matter
+            if worker.join().is_err() {
```

Comment:
Given this is quasi-async with the multi-threading, I think the only speed hit would be the array traversal being slowed down/being forced to be done linearly. If this invocation of slam needs perf improvements, this block can be measured.

If it needs improvements and we think async is the trick, a oneshot channel sender and the receivers could be turned into futures which are waited on in a `select!` (the `into_future` method in `StreamExt` turns it into a future which reports done when the other side of the channel dies).

Reply:
Yeah I suspect that the performance of this part just doesn't matter, because this is when the slam is ending. So by this point we should already have saturated the blockchain if we were able to do so, and saving some microseconds here won't make it a better test